src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -278,6 +326,17 @@ public void stop(Throwable cause) throws Exception {
          */
         private String cookie;
 
+        /**
+         * Needed for {@link BodyExecution#cancel}.
+         * {@code transient} because we cannot save a {@link BodyExecution} in {@link PlaceholderTask}:
+         * {@link ExecutorPickle} is written to the stream first, which holds a {@link PlaceholderTask},
+         * and the {@link BodyExecution} holds {@link PlaceholderTask.Callback} whose {@link WorkspaceList.Lease}
+         * is not processed by {@link WorkspaceListLeasePickle} since pickles are not recursive.
+         * So we make a best effort and only try to cancel a body within the current session.
```

Comment:
So IIUC, if you restart Jenkins while a Pipeline is executing inside of a `node` block, and then remove the Node from Jenkins, that's how you get to the "Agent foo was deleted, but do not have a node body to cancel" branch in RemovedNodeListener? Or `session` here refers to something else?

Reply:
Your understanding is correct. The patch to `DurableTaskStep` should still suffice to abort the build if it was inside `sh` when the node was removed.

If you remove the node _during_ the restart, a different bit of code in `ExecutorPickle` (existing long before this patch) kicks in and aborts the build before it even resumes the CPS VM thread.